File: tests/fixtures/go/circular_types.go
package fixtures

// Node represents a doubly linked circular list element.
type Node struct {
	ID    string
	Value interface{}
	Next  *Node
	Prev  *Node
}

// GraphNode represents a vertex in an adjacency graph with bidirectional edge references.
type GraphNode struct {
	ID          string
	Label       string
	Outgoing    []*GraphEdge
	Incoming    []*GraphEdge
	SelfPointer *GraphNode
}

// GraphEdge connects two GraphNodes with weight and properties.
type GraphEdge struct {
	ID     string
	Source *GraphNode
	Target *GraphNode
	Weight float64
}

// Scope represents an AST lexical scope with parent and child relationships.
type Scope struct {
	ScopeID     string
	ParentScope *Scope
	Children    []*Scope
	Symbols     map[string]*Symbol
}

// Symbol represents a variable or type binding in a Scope.
type Symbol struct {
	Name            string
	TypeSignature   string
	DeclaredInScope *Scope
	References      []*Scope
}

// BuildSampleDoublyLinkedList creates a sample cyclic 2-node structure.
func BuildSampleDoublyLinkedList() *Node {
	head := &Node{ID: "node_1", Value: "first"}
	tail := &Node{ID: "node_2", Value: "second"}

	head.Next = tail
	head.Prev = tail

	tail.Next = head
	tail.Prev = head

	return head
}

// ConnectGraphNodes links two GraphNode instances with a directed edge.
func ConnectGraphNodes(src, dst *GraphNode, edgeID string, weight float64) *GraphEdge {
	edge := &GraphEdge{
		ID:     edgeID,
		Source: src,
		Target: dst,
		Weight: weight,
	}
	src.Outgoing = append(src.Outgoing, edge)
	dst.Incoming = append(dst.Incoming, edge)
	return edge
}

File: tests/fixtures/typescript/circular_types.ts
/**
 * Mutually recursive and circular interface references for TypeScript.
 */

export interface TreeNode<T = string> {
    id: string;
    value: T;
    parent?: TreeNode<T>;
    children: Array<TreeNode<T>>;
    leftSibling?: TreeNode<T>;
    rightSibling?: TreeNode<T>;
}

export interface GraphNode {
    id: string;
    label: string;
    incomingEdges: Edge[];
    outgoingEdges: Edge[];
    metadata: Record<string, unknown>;
}

export interface Edge {
    id: string;
    source: GraphNode;
    target: GraphNode;
    weight: number;
    bidirectional: boolean;
}

export interface ScopeTree {
    scopeId: string;
    parentScope?: ScopeTree;
    childScopes: ScopeTree[];
    declaredSymbols: Map<string, SymbolBinding>;
}

export interface SymbolBinding {
    name: string;
    typeAnnotation: string;
    declaredInScope: ScopeTree;
    referencingScopes: ScopeTree[];
    shadowsParentSymbol?: SymbolBinding;
}

export function buildSampleGraph(): { root: GraphNode; edge: Edge } {
    const rootNode: GraphNode = {
        id: "node-1",
        label: "Root Node",
        incomingEdges: [],
        outgoingEdges: [],
        metadata: {},
    };

    const targetNode: GraphNode = {
        id: "node-2",
        label: "Child Node",
        incomingEdges: [],
        outgoingEdges: [],
        metadata: {},
    };

    const edge: Edge = {
        id: "edge-1-2",
        source: rootNode,
        target: targetNode,
        weight: 1.0,
        bidirectional: false,
    };

    rootNode.outgoingEdges.push(edge);
    targetNode.incomingEdges.push(edge);

    return { root: rootNode, edge };
}

export function traverseTreeDepthFirst<T>(node: TreeNode<T>, visitor: (item: T) => void): void {
    visitor(node.value);
    for (const child of node.children) {
        traverseTreeDepthFirst(child, visitor);
    }
}

export interface User {
    id: string;
    name: string;
    posts: Post[];
}

export interface Post {
    id: string;
    title: string;
    author: User;
    comments: Comment[];
}

export interface Comment {
    id: string;
    text: string;
    author: User;
    post: Post;
}

export function formatUser(user: User): string {
    return `${user.name} (${user.posts.length} posts)`;
}

